include/ginkgo/core/solver/chebyshev.hpp

```diff
@@ -0,0 +1,295 @@
+
+    /**
+     * Returns the solver operator used as the inner solver.
+     *
+     * @return the solver operator used as the inner solver
+     */
+    std::shared_ptr<const LinOp> get_solver() const { return solver_; }
+
+    /**
+     * Sets the solver operator used as the inner solver.
+     *
+     * @param new_solver  the new inner solver
+     */
+    void set_solver(std::shared_ptr<const LinOp> new_solver);
+
+    /**
+     * Copy-assigns a Chebyshev solver. Preserves the executor, shallow-copies
+     * inner solver, stopping criterion and system matrix. If the executors
+     * mismatch, clones inner solver, stopping criterion and system matrix onto
+     * this executor.
+     */
+    Chebyshev& operator=(const Chebyshev&);
+
+    /**
+     * Move-assigns a Chebyshev solver. Preserves the executor, moves inner
+     * solver, stopping criterion and system matrix. If the executors mismatch,
+     * clones inner solver, stopping criterion and system matrix onto this
+     * executor. The moved-from object is empty (0x0 and nullptr inner solver,
+     * stopping criterion and system matrix)
+     */
+    Chebyshev& operator=(Chebyshev&&);
+
+    /**
+     * Copy-constructs an Chebyshev solver. Inherits the executor,
+     * shallow-copies inner solver, stopping criterion and system matrix.
+     */
+    Chebyshev(const Chebyshev&);
+
+    /**
+     * Move-constructs an Chebyshev solver. Preserves the executor, moves inner
+     * solver, stopping criterion and system matrix. The moved-from object is
+     * empty (0x0 and nullptr inner solver, stopping criterion and system
+     * matrix)
+     */
+    Chebyshev(Chebyshev&&);
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Criterion factories.
+         */
+        std::vector<std::shared_ptr<const stop::CriterionFactory>>
+            GKO_FACTORY_PARAMETER_VECTOR(criteria, nullptr);
+
+        /**
+         * Inner solver factory.
+         */
+        std::shared_ptr<const LinOpFactory> GKO_FACTORY_PARAMETER_SCALAR(
+            solver, nullptr);
+
+        /**
+         * Already generated solver. If one is provided, the factory `solver`
+         * will be ignored.
+         */
+        std::shared_ptr<const LinOp> GKO_FACTORY_PARAMETER_SCALAR(
+            generated_solver, nullptr);
+
+        /**
+         * The pair of foci of ellipse. It is usually be {lower bound of eigval,
+         * upper bound of eigval} for real matrices.
+         */
+        std::pair<value_type, value_type> GKO_FACTORY_PARAMETER_VECTOR(
+            foci, value_type{0}, value_type{1});
+
+        /**
+         * Default initial guess mode. The available options are under
+         * initial_guess_mode.
+         */
+        initial_guess_mode GKO_FACTORY_PARAMETER_SCALAR(
+            default_initial_guess, initial_guess_mode::provided);
+
+        /**
+         * The number of scalar to keep
+         */
+        int GKO_FACTORY_PARAMETER_SCALAR(num_keep, 2);
```

Comment:
no, `num_keep` is to keep the generated scalar in the storage.
alpha and beta are changed by iteration, but they are fixed by the given bound.
It's used for fixed iteration runs because we do not need to refill these scalars for different apply

Reply:
But shouldn't we then just keep all scalars?

Reply:
when using it as normal solver, we may not have the maximum iteration information.
allocating one big dense matrix and then moving them to another one when full may not be a good approach.
we can also add the ability of workspace such that it can handle the std::vector<dense> then it should be more flexible for uncertain size.

Reply:
I think we can and should assume that this solver (which is mostly going to be used as a preconditioner anyway) has an iteration criterion. Then we can always get the max number of iteration, or throw if the iteration criterion is not there.
TBH I'm opposed to this factory parameter, because it's too much of an implementation detailed revealed to the user.

Reply:
That would also remove a bit of the complexity in the implementation.

Reply:
I can not assume that there's an iteration criterion. It can contain the residual norm or time criterion.
If we assume that, we should provide the standalone iteration parameter.
Users do not need to know the implementation details. 
The statement is to keep how many scalars for chebyshev to avoid the refill overhead.
I can introduce this usage later when we have the use case.
I think it should help the situation when kernel launch overhead is noticeable.

Reply:
You are right, this is very useful to negate the overhead. Still, I think we can assume that there is an iteration criterion somewhere in the combined one and use that. If not, we would just throw.

Reply:
You could set the default value to `unspecified`, which you define before. When the solver is generated, you can check if the parameter is `unspecified` and then try to extract an iteration criterion from the criteria. If that doesn't work, you throw with a message to either pass an iteration criterion or set this parameter to something else.
Or alternatively, just replace the criteria parameter with a `iterations` parameter and move the class into the preconditioner namespace. I think that is also a fine option, since that would be the main use case anyway.

Reply:
Thanks, I think it should be fine the way it is now.